### ISSUE TITLE
How to change complete pixi-spine object's opacity? And what's the right way to change slot images' opacity?

### ISSUE THREAD

**Issue by Friksel** [pixijs/pixi-spine#365]
Working my way through the pixi-spine API I'm now at the stage where I'd like to make attachments semi-transparent by Javascript.

In the Spine editor we would do this by adjusting the transparency-channel of the color of that attachment. In pixi-spine this seems to be represented by the `a` in the `attachment.color` object. However, when I change this color-`a`-value directly nothing happens. When changing the color values for `r`, `g` and/or `b` the color changes immediately though, without the need to call some `update()` method, so I was expecting changing the `a` value would do the same, but it's not.
I also saw a `tempColor` variable inside the attchment object, but changing the `a` on that doesn't change the visible opacity either.

I did the same test with changing the `a` of the `slot.color`, but I can't get that to work either.

At first I also tried setting the `opacity` of the pixi-spine object, but it didn't change the opacity. Than I tried setting the `opacity` of the attachment, but that didn't change it either. Also the `bone.opacity` didn't help (although that makes sense to me).

I'm sure I'm missing something here as I can't believe both the `a` value as the `opacity` values don't change the opacities.  But I don't know what I'm missing here. It doesn't seem to be that I'm missing calling an `update()` method as the colors update immediately.

I see now something that does change the opacity though: `slot.currentSprite.alpha = 0.5`. But I'm not sure if that's the way to go.

**Actually my questions are:**
1) How can we change the opacity of the full pixi-spine display object? [important]
2) How can we change the opacity of a single slot? [important]
     --> Should this be done with `slot.currentSprite.alpha = 0.5;`? It works, but what's the right way of doing this?

Thanks in advance!

**Comment by ivanpopelyshev** [comment]
> I see now something that does change the opacity though: `slot.currentSprite.alpha = 0.5`. But I'm not sure if that's the way to go.

Yeah, because slot container alpha is managed by `Spine.ts` file, and it overrides whatever you write in that param.

As for going inside core stuff: slot opacitym attachment opacity, sorry - cant help you with that. IF you find a good method, I'll add it to `examples` folder.

**Comment by ivanpopelyshev** [comment]
> It works, but what's the right way of doing this?

It doesnt exist in Spine API and pixi-spine recipes in `examples` folder, right? so there's no right method. Whatever you do - i publish - it will be new `right` method

**Comment by ivanpopelyshev** [comment]
Those questions are hard for me because there are tens of ways to do it, but i dont know which one will help most number of users or whether its understandable, or whether it will make a new bug.

**Comment by Friksel** [comment]
@ivanpopelyshev  

**slot and attachment colors and opacities**
With the Spine editor it's possible to set a color with alpha to the Slot and a color with alpha to the Attachment(s) that are inside that slot seperately. When both the slot AND the active attachment have a color/alpha assigned that's other than white, the Spine editor uses a filter to calculate the resulting color/alpha. If the slot is full Red only (so full red, rgb = 255, 0, 0) and the Region Attachment is full green only (so full green, rgb = 0, 255, 0), the result is that the attachment is displayed black (as the attachment only passes through the green, but all green was filtered from the upper lying slot, as that was only letting through red).

I think it's wise to stay with that logic in pixi-spine; so Slot can have colors and alpha, and attachment can have their own colors and alpha. And use the same calculation to calculate the resulting color of the two.

**total skeleton color and alpha**
In the Spine editor AFAIK it isn't possible to color the full skeleton and set it's alpha for the full skeleton. But it would make a lot of sense to me to be able to do this in pixi-spine by code though, as it's a Pixi displayobject and I would expect the displayobject's `alpha` variable to work just like other Pixi displayobjects.

Making a pixi-spine skeleton object semi transparent, or color it, is something with nice use cases. Like a walking ghost that you want to make half invisible by code. Or when something happens by game logic, or a user interaction. The same with being able to change the colors of the full skeleton. 

But unfortunately when changing the alpha of the full skeleton it's not enough the change all seperate images's opacities. As when these overlapping images get semi transparent we get undesirable effects as they are now see-though and could not only reveal parts we don't want to reveal, but also it doesn't look great and as planned. So the alpha for the full object should actually only be applied to the 'container' of all the slots and attachments in a way that only the container gets semi transparent by that action. Hope I explained this well.

AFAIK everywhere in Pixi the parameter for transparancy is called `alpha` and the parameter for tinting (changing color) is called `tint`. So if you'd ask me I'd stick with these terminology wherever possible as we're in the Pixi environment. But if Spine runtimes are talking about transparency on some places it's probably easer to have both, as these names don't conflict. So people thinking 'Pixi' could find `alpha` and `tint` as expected, and people with Spine in their mind can find the Spine equivalents as found in the Spine API, if any.

**If you'd ask me this is what I would prefer:**
``` javascript
const mySpineObj = new PIXI.spine.Spine(this.loader.resources.boxrotator.spineData);
const slot = this.boxrotator.skeleton.slots[0];

mySpineObj.skeleton.alpha = 0.5; // SET THE ALPHA OF THE FULL SKELETON
mySpineObj.skeleton.tint = 0xFF0000; // SET TINT OF FULL SKELETON TO RED

slot.alpha = 0.5; // BEING ABLE TO SET THE SLOT'S OPACITY DIRECTLY ON THE SLOT. INFLUENCES ALL UNDERLYING ATTACHMENTS
slot.tint = 0xFF0000; // SET TINT OF SLOT TO RED

slot.attachment.alpha = 0.5; // CONVENIENCE REFERENCE: BEING ABLE TO SET THE SLOT'S CURRENTLY SELECTED ATTACHMENT'S OPACITY (THE OPACITY HERE COULD PERHAPS BE A DIRECT REFERENCE TO slot.currentSprite.alpha?)
slot.attachment.tint = 0xFF0000; // SET TINT OF ATTACHMENT TO RED
```

Guess with this we got the `slot.attachment.alpha` in the lib now as `slot.currentSprite.alpha`, so it should be too hard to make a reference for convenience. For the others I'm still searching if I can find it in `pixi-spine`, but it doesn't seem to be there.

Hope this makes sense and helps!
